pdns/iputils.hh

```diff
@@ -369,6 +370,313 @@ private:
+    // it is easier to copy the nodes than tree.
+    // also acts as handy compactor
+    for(auto node: rhs._nodes)
+      insert(node->first).second = node->second;
+  }
+
+  NetmaskTree& operator=(const NetmaskTree& rhs) {
+    clear();
+    // see above.
+    for(auto node: rhs._nodes)
+      insert(node->first).second = node->second;
+    return *this;
+  }
+
+  //<! Assigment operator for value-pair, throws if not found
+  const value_type& operator[](const key_type& rhs) const {
+    const node_type *value = lookup(rhs.getNetwork(), rhs.getBits());
+    if (value == nullptr) throw std::range_error(rhs.toString() + string(" not found"));
+    return value->second;
+  }
+
+  //<! Assigment operator for value-pair, creates new if not found
+  T& operator[](const Netmask& rhs) {
+    return insert(rhs).second;
+  }
+
+  const typename std::vector<node_type*>::const_iterator begin() const { return _nodes.begin(); }
+  const typename std::vector<node_type*>::const_iterator end() const { return _nodes.end(); }
+
+  typename std::vector<node_type*>::iterator begin() { return _nodes.begin(); }
+  typename std::vector<node_type*>::iterator end() { return _nodes.end(); }
+
+  node_type& insert(const string &mask) {
+    return insert(key_type(mask));
+  }
+
+  //<! Creates new value-pair in tree and returns it.
+  node_type& insert(const key_type& key) {
+    // lazily initialize tree on first insert.
+    if (!root) root = unique_ptr<TreeNode>(new TreeNode(0));
+    TreeNode* node = root.get();
+    node_type* value = nullptr;
+
+    if (key.getNetwork().sin4.sin_family == AF_INET) {
+      std::bitset<32> addr(be32toh(key.getNetwork().sin4.sin_addr.s_addr));
+      int bits = 0;
+      // we turn left on 0 and right on 1
+      while(bits < key.getBits()) {
+        uint8_t val = addr[31-bits];
+        if (val)
+          node = node->make_right();
+        else
+          node = node->make_left();
+        bits++;
+      }
+      // only create node if not yet assigned
+      if (!node->node4) node->node4 = unique_ptr<node_type>(new node_type());
+      value = node->node4.get();
+    } else {
+      uint64_t* addr = (uint64_t*)key.getNetwork().sin6.sin6_addr.s6_addr;
+      std::bitset<64> addr_low(be64toh(addr[1]));
+      std::bitset<64> addr_high(be64toh(addr[0]));
+      int bits = 0;
+      while(bits < key.getBits()) {
+        uint8_t val;
+        // we use high address until we are
+        if (bits < 64) val = addr_high[63-bits];
+        // past 64 bits, and start using low address
+        else val = addr_low[127-bits];
+
+        // we turn left on 0 and right on 1
+        if (val)
+          node = node->make_right();
+        else
+          node = node->make_left();
+        bits++;
+      }
+      // only create node if not yet assigned
+      if (!node->node6) node->node6 = unique_ptr<node_type>(new node_type());
+      value = node->node6.get();
+    }
+    // assign key
+    value->first = key;
```

Comment:
Maybe the key initialize could be moved into the `node_type` creation: `new node_type(key, value_type{})`.